api/provisioning.go

```diff
@@ -1800,64 +1759,70 @@ func checkConfigMapping(applied []Configlet, newconfiglets []Configlet) (bool,
+
+func splitToConfigletAndBuilder(configlets []Configlet) ([]string, []string, []string, []string, error) {
+
 	var configletNames []string
-	var builderNames []string
-	var rmConfigletNames []string
-	var rmBuilderNames []string
+	var configletKeys []string
 
-	configletMap := make(map[string]string)
+	var builderNames []string
+	var builderKeys []string
 
-	for _, configlet := range rmConfiglets {
+	for _, configlet := range configlets {
 		switch configlet.Type {
-		case "Static":
-			fallthrough
-		case "Generated":
-			fallthrough
-		case "Reconciled":
-			rmConfigletNames = append(rmConfigletNames, configlet.Name)
+		case "Static", "Generated", "Reconciled":
+			configletNames = append(configletNames, configlet.Name)
+			configletKeys = append(configletKeys, configlet.Key)
+
 		case "Builder":
-			rmBuilderNames = append(rmBuilderNames, configlet.Name)
+			builderNames = append(builderNames, configlet.Name)
+			builderKeys = append(builderKeys, configlet.Key)
+
 		default:
-			return false, nil, nil, nil, nil,
+			return nil, nil, nil, nil,
 				errors.Errorf("Configlet [%s] Invalid Type [%s]", configlet.Name, configlet.Type)
 		}
-		configletMap[configlet.Key] = configlet.Name
+	}
+
+	return configletKeys, configletNames, builderKeys, builderNames, nil
+}
+
+
+// checkRemoveConfigMapping Creates map of configlets that needs to be there after removal of
+// specific configlets
+func checkRemoveConfigMapping(applied []Configlet, rmConfiglets []Configlet) (bool,
+	[]string, []string, []string, []string,[]string, []string, []string, []string, error) {
+
+	rmConfigletKeys, rmConfigletNames, rmBuilderKeys, rmBuilderNames, err := splitToConfigletAndBuilder(rmConfiglets)
+	if err != nil {
+		return false, nil,nil,nil,nil,nil, nil, nil, nil, err
+	}
+
+	rmKeys := map[string]Configlet{}
+	for _, c := range rmConfiglets {
+		rmKeys[c.Key] = c
 	}
 
 	var actionReqd bool
-	for _, configlet := range applied {
-		// don't process already processed configlets
-		if _, found := configletMap[configlet.Key]; found {
-			continue
+	var configletsToRemain []Configlet
+	for _, c := range applied  {
+		if _, found := rmKeys[c.Key]; ! found {
+			configletsToRemain = append(configletsToRemain, c)
 		}
 
-		// didn't find this configlet in either maps, so it's new
 		actionReqd = true
+	}
 
-		switch configlet.Type {
-		case "Static":
-			fallthrough
-		case "Generated":
-			fallthrough
-		case "Reconciled":
-			configletNames = append(configletNames, configlet.Name)
-		case "Builder":
-			builderNames = append(builderNames, configlet.Name)
-		default:
-			return false, nil, nil, nil, nil,
-				errors.Errorf("Configlet [%s] Invalid Type [%s]", configlet.Name, configlet.Type)
-		}
+	configletKeys, configletNames, builderKeys, builderNames, err2 := splitToConfigletAndBuilder(configletsToRemain)
+	if err2 != nil {
+		return false, nil, nil, nil, nil, nil, nil,nil, nil, err2
 	}
 
 	if len(configletNames) == 0 && len(builderNames) == 0 {
```

Comment:
@cheynearista personally I don't get this special case of both being length 0, what was/is this supposed to solve? Maybe I can add it as a comment once I know the rationale

Reply:
@hansbogert I think this was to handle the situation where applied []Configlet matches rmConfiglets []Configlet (i.e. User is removing all configlets that are currently applied). In this case actionReqd would need to be true so the calling func can create the 'action' to removed them.

Reply:
Then my `actionreqd = true` doesn't make much sense.
Would it by equivalent if we do a check for the following:

`actionReqd = len(builderNames) + len(configletNames) != len(applied)`

?

Reply:
I think that would make sense if after line [1810](https://github.com/aristanetworks/go-cvprac/pull/70/files#diff-af964366e44ed2bd45c6078a428dc200R1810) you added a `continue`.  `actionReqd` would need to be set to true if `configletsToRemain` has something different than what `applied` was passed in with (i.e. There were applied configlets removed).

Reply:
I wasn't sure if we partially agreed or not at all, I've redid the logic as per my own suggestion, please review if this logic is sound.

Reply:
@cheynearista gentle nudge :relaxed: